### DISCUSSION TITLE
Block restrict mass_fraction_vars variable

### DISCUSSION THREAD

**Discussion by 1runer** [idaholab/moose#26590] [Q&A Modules: Porous Flow] [answered]
Hi all, hi @cpgr,

Is there any way to block restrict (and solve) the mass_fraction variable to a particular block?
Or a different approach to achieve this?

I saw this [post](https://github.com/idaholab/moose/discussions/20139). This doesn't solve my question (?) since I want to have PorousFlow everywhere active but with different amount of components? 

Exemplary problem setting:
Having a 2D fracture embedded in a 3D matrix where a tracer is supposed to just flow along the fracture. 
The tracer can't diffuse into the matrix (e.g. due to pore size limitations). Unfortunately, numerical diffusion causes always some tracer flow perpendicular to the fracture. Therefore, I would like to restrict the mass_fraction_variable tracer to just the fracture block.

Unfortunately, PorousFlow doesn't support the use of two dictators (one for the fracture with 1p2c and one for the matrix with 1p1c) and two PorousFlowMassFraction materials.
```
[UserObjects]
  [dictator1]
    type = PorousFlowDictator
    porous_flow_vars = 'pressure tracer'
    number_fluid_phases = 1
    number_fluid_components = 2
    block = fracture
  []
  [dictator2]
    type = PorousFlowDictator
    porous_flow_vars = 'pressure'
    number_fluid_phases = 1
    number_fluid_components = 1
    block = 'matrix'
  []
[]

[Materials]
  [massfrac1]
    type = PorousFlowMassFraction
    mass_fraction_vars = 'tracer'
    block = 'fracture'
  []
  [massfrac2]
    type = PorousFlowMassFraction
    block = 'matrix'
  []
[]
```
Errors with:
```*** ERROR ***
The following parameter error occurred in the object "massfrac2", of type "PorousFlowMassFraction".

mass_fraction_vars:
    This value must be equal to the Dictator's num_phases (1) multiplied by num_components-1 (1)
```

**Comment by GiudGiud** [comment]
>  since I want to have PorousFlow everywhere active but with different amount of components?

I think you will need duplicate the entire input (every UOs etc) to vary that amount of components.
Otherwise did you try having the same number of components everywhere but some naturally stay at 0 over part of the domain?

**Reply by 1runer** [reply]
I don't have other UOs involved besides FluidProperties (SimpleFluid). 
Ok, I was able to solve the error by overwriting the GlobalParam statement that I had forgotten initially (my bad).

But still doesn't work as I would like to have it. 
The simulations fails due to FACTOR_OUTMEMORY. Same applies for just block-restricting tracer-related kernels to the fracture.

```
  [massfrac1
    type = PorousFlowMassFraction
    mass_fraction_vars = 'tracer'
    block = 'fracture'
    PorousFlowDictator = dictator1
  []
  [massfrac2]
    type = PorousFlowMassFraction
    block = 'matrix'
    PorousFlowDictator = dictator2
  []
```
```
 0 Nonlinear |R| = 8.063257e-01
    |residual|_2 of individual variables:
                   pressure: 0.806326
                   tracer: 2.14625e-17
      0 Linear |R| = 4.732462e+02
  Linear solve did not converge due to DIVERGED_PC_FAILED iterations 0
                 PC failed due to FACTOR_OUTMEMORY 
Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 0
 Solve Did NOT Converge!
Currently Executing
  Finished Solving                                                                       
Aborting as solve did not converge
```
Adding a NullKernel for 'tracer' at least solves. But with overshooting/undershooting in most areas, mostly in the matrix.

Option 2 would be something like you mentioned:
```
[Kernels]
  [mass0]
    type = PorousFlowMassTimeDerivative
    fluid_component = 1
    variable = pressure
  []
  [adv0]
    type = PorousFlowFullySaturatedAdvectiveFlux
    fluid_component = 1
    variable = pressure
  []
  [adv1]
    type = PorousFlowFullySaturatedAdvectiveFlux 
    fluid_component = 0
    variable = tracer
    block = fracture
  []
  [mass1]
    type = PorousFlowMassTimeDerivative
    fluid_component = 0
    variable = tracer
  []
  [disp1]
    type = PorousFlowDispersiveFlux
    # Diffusion also 0
    fluid_component = 0
    variable = tracer
    disp_trans = 0
    disp_long = 0
  []
[]
```

This solves smoothly. 
But for larger meshes/timestepping this will produce artificial diffusion into the matrix?
The fracture aperture and therefore the mass injected into the 2D fracture is tiny (1e-4 m) compared to the surrounding mesh size.

**Reply by GiudGiud** [reply]
If a null kernel helps things you need to fix your block restriction of variables.

Set the troubleshooting failed solves help page for more info on solving

**Reply by 1runer** [reply, marked as answer]
I managed to block-restrict the mass_fraction variable 'tracer' by adding the dictator manually to each PorousFlow-related section and duplicating lot's of code.

The results for a low-permeable matrix (1e-20 m2) are now identically to the results of a pure 2D model without any 3D matrix. 
In contrast just limiting advection to the fracture and setting the diffusion_coef to 0 in the matrix does slow down the tracer propagation far below expected values even if pressure, pressure gradient and velocity are identical in all three cases: 
[LINK to the INPUT FILES](https://github.com/1runer/porous_flow_stuff)

![tracer_comp](https://github.com/idaholab/moose/assets/26140953/d166cd6d-a0b7-4b45-b1ca-01e1c319dd07)

**Reply by cpgr** [reply]
Interesting! By 'duplicating lots of code' do you just mean in the input file, or did you have to modify the source code?

**Reply by 1runer** [reply]
I just had to modify the input file and adjust different Kernels, Materials, Auxkernels, etc. to the different Dictators (1 with pressure and tracer or 2 with pressure only)
No modification in the source code.
Solve is perfect as with one Dictator. :)

**Reply by cpgr** [reply]
This is really useful to know!
